src/Microsoft.ML.Data/Transforms/CopyColumnsTransform.cs

```diff
@@ -47,120 +136,218 @@ public sealed class Arguments : TransformInputBase
+            // for each added column
+            //   string: output column name
+            //   string: input column name
+
+            var length = ctx.Reader.ReadInt32();
+            var columns = new (string Source, string Name)[length];
+            for (int i = 0; i < length; i++)
+            {
+                columns[i].Name = ctx.LoadNonEmptyString();
+                columns[i].Source = ctx.LoadNonEmptyString();
+            }
+            return new CopyColumnsTransformer(env, columns);
+        }
+
+        public static IDataTransform Create(IHostEnvironment env, Arguments args, IDataView input)
         {
+            Contracts.CheckValue(env, nameof(env));
+            env.CheckValue(args, nameof(args));
+            var transformer = new CopyColumnsTransformer(env, args.Column.Select(x => (x.Source, x.Name)).ToArray());
+            return transformer.CreateRowToRowMapper(input);
         }
 
-        public CopyColumnsTransform(IHostEnvironment env, Arguments args, IDataView input)
-            : base(env, RegistrationName, env.CheckRef(args, nameof(args)).Column, input, null)
+        public ISchema GetOutputSchema(ISchema inputSchema)
         {
-            Host.AssertNonEmpty(Infos);
-            Host.Assert(Infos.Length == Utils.Size(args.Column));
-            SetMetadata();
+            _host.CheckValue(inputSchema, nameof(inputSchema));
+            // Validate schema.
+            return Transform(new EmptyDataView(_host, inputSchema)).Schema;
         }
 
-        private CopyColumnsTransform(IHost host, ModelLoadContext ctx, IDataView input)
-            : base(host, ctx, input, null)
+        public void Save(ModelSaveContext ctx)
         {
-            Host.AssertValue(ctx);
+            _host.CheckValue(ctx, nameof(ctx));
+            ctx.CheckAtModel();
+            ctx.SetVersionInfo(GetVersionInfo());
 
             // *** Binary format ***
-            // <base>
+            // int: number of added columns
+            // for each added column
+            //   string: output column name
+            //   string: input column name
+            ctx.Writer.Write(_columns.Length);
+            foreach (var column in _columns)
+            {
+                ctx.SaveNonEmptyString(column.Name);
+                ctx.SaveNonEmptyString(column.Source);
+            }
+        }
 
-            Host.AssertNonEmpty(Infos);
-            SetMetadata();
+        private RowToRowMapperTransform CreateRowToRowMapper(IDataView input)
+        {
+            var mapper = new CopyColumnsRowMapper(_host, input.Schema, _columns);
+            return new RowToRowMapperTransform(_host, input, mapper);
+        }
+
+        public IDataView Transform(IDataView input)
+        {
+            return CreateRowToRowMapper(input);
+        }
+    }
+
+    internal class CopyColumnsRowMapper : IRowMapper
+    {
+        private readonly ISchema _schema;
+        private readonly Dictionary<int, int> _colMap;
```

Comment:
>_colMap [](start = 46, length = 7)

map what to what? Old to new or new to old?

Reply:
Rename to be clearer

---
In reply to: [212479904](https://github.com/dotnet/machinelearning/pull/706#discussion_r212479904) [](ancestors = 212479904)